speeduino.ino

```diff
@@ -199,49 +261,49 @@ void setup()
   digitalWrite(pinTPS, LOW);
   
   //Calculate the number of degrees between cylinders
+  // FIXME these crank angles should be configurable via tunerstudio
+  int angles[4] = { 0, 0, 0, 0};
   switch (configPage1.nCylinders) {
     case 1:
-      channel1Degrees = 0;
+      angles[0] = 0;
       break;
-    case 2:
-      channel1Degrees = 0;
-      channel2Degrees = 180;
+    case 2: // this is probably only correct for a 2 cyl straight engine, not a V2 like a Harley or a boxer
+      angles[0] = 0;
+      angles[1] = 180;
       break;
-    case 3:
-      channel1Degrees = 0;
-      channel2Degrees = 120;
-      channel3Degrees = 240;
+    case 3: // this is only correct for firing order 132
+      angles[0] = 0;
+      angles[1] = 480; // 120deg when non sequential;
+      angles[2] = 240;
       break;
     case 4:
-      channel1Degrees = 0;
-      channel2Degrees = 180;
-      break;
-    case 6:
-      channel1Degrees = 0;
-      channel2Degrees = 120;
-      channel3Degrees = 240;
-      break;
-    default: //Handle this better!!!
-      channel1Degrees = 0;
-      channel2Degrees = 180;
+      // falltrough on purpose.
+    default: // this is only correct for 4 cyl straight engine, firing order 1342
+      angles[0] = 0;
+      angles[1] = 540; // 540 deg from TDC on cyl1
+      angles[2] = 180;
+      angles[3] = 360;   // 360 deg from TDC on cyl1
       break;
   }
+
+  // set up
+  channels.setCylCrankAngles(configPage1.nCylinders, angles);
 }
 
 void loop() 
   {
-      mainLoopCount++;    
-      //Check for any requets from serial. Serial operations are checked under 2 scenarios:
-      // 1) Every 64 loops (64 Is more than fast enough for TunerStudio). This function is equivalent to ((loopCount % 64) == 1) but is considerably faster due to not using the mod or division operations
-      // 2) If the amount of data in the serial buffer is greater than a set threhold (See globals.h). This is to avoid serial buffer overflow when large amounts of data is being sent
-      if ( ((mainLoopCount & 63) == 1) or (Serial.available() > SERIAL_BUFFER_THRESHOLD) ) 
```

Comment:
What is the intent of removing this line?

The serial buffer check is there to ensure that, if the serial receive buffer is starting to fill up, that the command() function will be called, even if it's earlier than the normal 64 loops. WIthout this check, I was seeing corruption of serial data when large blocks of data (ie whole tables) were being sent by TS all at once and were overflowing the serial buffer before the 64 loops were up.